### PR TITLE
EIP multi NIC: During Repair(), ignore IPv6 NAT table error

### PR DESCRIPTION
**Pull request by martinkennelly**
See commits for more details.

### CODE REVIEW

**Comment by martinkennelly** [comment]
This error is not seen downstream in OCP because the necessary kernel mods are available but on certain distros they are not available by default.

**Comment by martinkennelly** [comment]
Test failure:
```
2024-02-27T10:44:58.2952555Z --- FAIL: TestNetworkPolicyV2Conformance (312.36s)
2024-02-27T10:44:58.2957147Z     --- FAIL: TestNetworkPolicyV2Conformance/AdminNetworkPolicyEgressSCTP (10.36s)
2024-02-27T10:44:58.2960048Z         --- PASS: TestNetworkPolicyV2Conformance/AdminNetworkPolicyEgressSCTP/Should_support_an_'allow-egress'_policy_for_SCTP_protocol;_ensure_rule_ordering_is_respected (0.18s)
2024-02-27T10:44:58.2963456Z         --- PASS: TestNetworkPolicyV2Conformance/AdminNetworkPolicyEgressSCTP/Should_support_an_'allow-egress'_policy_for_SCTP_protocol_at_the_specified_port (3.16s)
2024-02-27T10:44:58.2966699Z         --- FAIL: TestNetworkPolicyV2Conformance/AdminNetworkPolicyEgressSCTP/Should_support_an_'deny-egress'_policy_for_SCTP_protocol;_ensure_rule_ordering_is_respected (3.51s)
2024-02-27T10:44:58.2970005Z         --- PASS: TestNetworkPolicyV2Conformance/AdminNetworkPolicyEgressSCTP/Should_support_a_'deny-egress'_policy_for_SCTP_protocol_at_the_specified_port (3.15s)
2024-02-27T10:44:58.2973861Z         --- PASS: TestNetworkPolicyV2Conformance/AdminNetworkPolicyEgressSCTP/Should_support_an_'pass-egress'_policy_for_SCTP_protocol;_ensure_rule_ordering_is_respected (0.17s)
2024-02-27T10:44:58.2977095Z         --- PASS: TestNetworkPolicyV2Conformance/AdminNetworkPolicyEgressSCTP/Should_support_a_'pass-egress'_policy_for_SCTP_protocol_at_the_specified_port (0.17s)
```

Unrelated to this PR.

**Comment by martinkennelly** [comment]
Flake issue here: https://github.com/ovn-org/ovn-kubernetes/issues/4142

**Comment by martinkennelly** [comment]
The reason we check both IP families during `Repair()`, is because of IP conversion.

**Comment by martinkennelly** [comment]
I didn't try to detect if the NAT table is present via a mechanism for iptables legacy and nft because its a whack-a-mole for maintainability.

The is an existing retry mechanism to cover if iptables lock is held.

**Review by flavio-fernandes — COMMENTED** [review]
As discussed here: https://github.com/ovn-org/ovn-kubernetes/pull/3749/files#r1497791405

```
❯ cat /etc/*release* | head -1
Fedora release 37 (Thirty Seven)
❯ lsmod | grep -c ip6table_nat
0
```

Before:

```
F0227 19:43:08.499676    4072 ovnkube.go:136] failed to run ovnkube: failed to start node network controller: failed to start default node network controller: failed to run egress IP controller: failed to run EgressIP controller because repairing node failed: failed to list IPTable IPv4 rules: failed to retrieve iptables table nat: exit status 1  

^^----  it is not really IPv4; typo in the code, which is also being fixed in this PR
```

After applying this PR:

```
❯ git checkout martin/eip-ignore-ipv6-if-no-kernmod
❯ git rev-parse HEAD
e9cbac994581c30701531efd104aff2b24682367
```

It is working better now, but I did not manage to see the message `klog.Warningf("Failed to list IPTable IPv6 rules: %v", err)` in logs. Did you?

**Comment by freedge** [comment]
had to use this patch on kind as it seems to detect the wrong ip6tables alternative
https://github.com/kubernetes-sigs/kind/blob/7c3e01fc1c25d833dd9938789a8f26b184472e8f/images/base/files/usr/local/bin/entrypoint#L415

**Comment by jcaamano** [comment]
@martinkennelly seeking to clarify the statement here.

Basically any iptables nat error for ipv6 in `repairNode` other than a resource error (exit status 4) is going to be interpreted as ipv6  iptables nat not available for the platform and continue on. I think it is more-less fine as there would be nothing stale to remove after boot so the only other possible cause is someone unloading the kernel module explicitly which which would be unexpected to me.

The potential issue here is hiding other type of errors (ErrExecutableNotFound for example). This is what I get when I try manually
```
root@ovn-worker2:/# ip6tables -t nat -L 1> /dev/null; echo $?
ip6tables v1.8.7 (legacy): can't initialize ip6tables table `nat': Table does not exist (do you need to insmod?)
Perhaps ip6tables or your kernel needs to be upgraded.
3
```
Notice that the error goes to stderr and exit status looks 3 to me instead of 1?

Looking at
https://github.com/ovn-org/ovn-kubernetes/blob/b4388c5a8766e35d5ae5d63833fd7ee00cf0592f/go-controller/pkg/node/iptables/iptables_manager.go#L212-L217
You might want to add buf to the error that you return so we get those error details.

**Comment by jcaamano** [comment]
welp, it's ip6tables-save giving a different exit status
```
root@ovn-worker:/# ip6tables-save -t nat; echo $?
ip6tables-save v1.8.7 (legacy): Cannot initialize: iptables who? (do you need to insmod?)

1
```

**Review by jcaamano — APPROVED** [review]

**Review by trozet — APPROVED** [review]

File: go-controller/pkg/node/iptables/iptables_manager.go [hunk 1]
```diff
@@ -211,7 +211,7 @@ func getChainRuleArgs(ipt iptables.Interface, table iptables.Table, chain iptabl
 	buf := bytes.NewBuffer(nil)
 	err := execIPTablesWithRetry(func() error {
 		if err := ipt.SaveInto(table, buf); err != nil {
-			return fmt.Errorf("failed to retrieve iptables table %s: %v", table, err)
+			return fmt.Errorf("failed to retrieve iptables table %s: %v: %s", table, err, buf)
```
Comment (flavio-fernandes) [inline go-controller/pkg/node/iptables/iptables_manager.go:214]: nice!

**Pull request merged**